File: app/src/main/java/com/stevensekler/android/movies/Fragments/BackdropAdapter.java
package com.stevensekler.android.movies.Fragments;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;

import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;
import com.stevensekler.android.movies.DetailActivity;
import com.stevensekler.android.movies.R;

import java.util.ArrayList;

/**
 * Created by Szekely Istvan on 8/27/17.
 *
 */

public class BackdropAdapter extends RecyclerView.Adapter<BackdropAdapter.BackdropViewHolder> {

    private ArrayList<String> images;
    private Context context;


    public BackdropAdapter(ArrayList<String> images, Context context) {
        this.images = images;
        this.context = context;
    }

    @Override
    public BackdropAdapter.BackdropViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(context);
        View view = inflater.inflate(R.layout.backdrop_recycler_view, parent, false);
        return new BackdropAdapter.BackdropViewHolder(view);
    }


    @Override
    public int getItemCount() {
        return images.size();
    }

    @Override
    public void onBindViewHolder(BackdropAdapter.BackdropViewHolder holder, int position) {

        Picasso.with(context)
                .load(images.get(position))
                .placeholder(R.drawable.blank_landscape92)
                .into(holder.backdropImage);
    }

    class BackdropViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        ImageView backdropImage;
        ProgressBar imageLoading;

        public BackdropViewHolder(final View itemView) {
            super(itemView);
            itemView.setOnClickListener(this);
            backdropImage = itemView.findViewById(R.id.backdrop_image);
        }

        @Override
        public void onClick(View view) {
            int position = getAdapterPosition();
            String s = images.get(position).replace("w92", "w500");
            ImageView mainImage = (ImageView) ((DetailActivity) context).findViewById(R.id.imageView2);
            imageLoading = (ProgressBar) ((DetailActivity) context).findViewById(R.id.backdrop_image_loading);
            imageLoading.setVisibility(View.VISIBLE);
            Picasso.with(context)
                    .load(s)
                    .noPlaceholder()
                    .into(mainImage, new Callback() {
                        @Override
                        public void onSuccess() {
                            imageLoading.setVisibility(View.GONE);
                        }

                        @Override
                        public void onError() {
                            imageLoading.setVisibility(View.GONE);
                        }
                    });
        }
    }
}

File: app/src/main/java/com/stevensekler/android/movies/Fragments/BackdropAdapter.kt
package com.stevensekler.android.movies.Fragments

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar

import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso
import com.stevensekler.android.movies.DetailActivity
import com.stevensekler.android.movies.R

import java.util.ArrayList

/**
 * Created by Szekely Istvan on 8/27/17.
 *
 */

class BackdropAdapter(private val images: ArrayList<String>, private val context: Context) : RecyclerView.Adapter<BackdropAdapter.BackdropViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BackdropAdapter.BackdropViewHolder {
        val inflater = LayoutInflater.from(context)
        val view = inflater.inflate(R.layout.backdrop_recycler_view, parent, false)
        return BackdropViewHolder(view)
    }


    override fun getItemCount(): Int {
        return images.size
    }

    override fun onBindViewHolder(holder: BackdropAdapter.BackdropViewHolder, position: Int) {

        Picasso.with(context)
                .load(images[position])
                .placeholder(R.drawable.blank_landscape92)
                .into(holder.backdropImage)
    }

    inner class BackdropViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        var backdropImage: ImageView
        lateinit var imageLoading: ProgressBar

        init {
            itemView.setOnClickListener(this)
            backdropImage = itemView.findViewById(R.id.backdrop_image)
        }

        override fun onClick(view: View) {
            val position = adapterPosition
            val s = images[position].replace("w92", "w500")
            val mainImage = (context as DetailActivity).findViewById<View>(R.id.imageView2) as ImageView
            imageLoading = context.findViewById<View>(R.id.backdrop_image_loading) as ProgressBar
            imageLoading.visibility = View.VISIBLE
            Picasso.with(context)
                    .load(s)
                    .noPlaceholder()
                    .into(mainImage, object : Callback {
                        override fun onSuccess() {
                            imageLoading.visibility = View.GONE
                        }

                        override fun onError() {
                            imageLoading.visibility = View.GONE
                        }
                    })
        }
    }
}
